src/lib.rs

```diff
@@ -553,6 +606,133 @@ impl CompileOptions {
         }
     }
 
+    /// Sets the callback for handling the `#include` directive.
+    ///
+    /// The arguments to the callback are the name of the source being requested,
+    /// the type of include directive (`Relative` for `#include "foo"`, `Standard` for `#include <foo>`),
+    /// the name of the source containing the directive and the current include depth from the original
+    /// source.
+    ///
+    /// The return value of the callback should be `Ok` if the source was successfully found,
+    /// and an `Err` containing some suitable error message to display otherwise.
+    ///
+    /// Behaviour note: If `Err` is returned for a `Relative` include request, the callback will be
+    /// tried again with `Standard`, which is similar to include directive behaviour in C.
+    pub fn set_include_callback<F>(&mut self, f: F)
+        where F : Fn(&str, IncludeType, &str, usize) -> result::Result<ResolvedInclude, String> + 'a
+    {
+        use std::ffi::{CString, CStr};
+        use std::mem;
+
+        let f = Box::new(f);
+        let f_ptr = &*f as *const F;
+        self.f = Some(f as Box<Fn(&str, IncludeType, &str, usize) -> result::Result<ResolvedInclude, String> + 'a>);
+        unsafe {
+            ffi::shaderc_compile_options_set_include_callbacks(self.raw, resolver::<'a, F>, releaser, f_ptr as *const c_void as *mut c_void);
+        }
+
+        struct OkResultWrapper {
+            source_name: CString,
+            content: CString,
+            wrapped: ffi::shaderc_include_result
+        }
+
+        struct ErrResultWrapper {
+            error_message: CString,
+            wrapped: ffi::shaderc_include_result
+        }
+
+        extern "C" fn resolver<'a, F>(user_data: *mut c_void,
+                               requested_source: *const c_char,
+                               type_: c_int,
+                               requesting_source: *const c_char,
+                               include_depth: size_t) -> *mut ffi::shaderc_include_result
+            where F : Fn(&str, IncludeType, &str, usize) -> result::Result<ResolvedInclude, String> + 'a
+        {
+            let result = panic::catch_unwind(move || {
+                let f = unsafe { &*(user_data as *const F) };
+                let requested_source = unsafe { CStr::from_ptr(requested_source).to_string_lossy() };
+                let type_ = match type_ {
+                    0 => IncludeType::Relative,
+                    1 => IncludeType::Standard,
+                    x => panic!("Unknown Include Type returned from libshaderc: {}", x)
```

Comment:
We typically prefer all lowercase words (except known uppercase terms) in error messages in compilers. For example, clang diagnostics: https://github.com/llvm-mirror/clang/blob/master/include/clang/Basic/DiagnosticSemaKinds.td

Also applies to the following error messages.

Reply:
Just to be clear, those error messages aren't compiler output though - they're panics, and they reflect the fact that they're highlighting errors due to callback code, not due to syntactic/semantic problems in the shader source. I see that the panics emitted by the standard library also follow the lowercase convention, so I'll switch to that anyway.